proposals/0034-glulx-random.md

```diff
@@ -0,0 +1,134 @@
+    which it currently accomplishes through Glulxe's `--rngseed` argument. So
+    rather than seeding Inform's new RNG with a timestamp, we should instead seed
+    it with `@random`. Even though the interpreter's RNG may be flawed, one call
+    with a high enough range is probably much safer than calling it with a range
+    of only 4 or 8. It has not yet been determined whether a full 32-bit `@random`
+    call is safe in affected interpreters, or instead whether it should make a
+    ranged call using a high prime number, or if any priming is necessary.
+
+3. We need to replace the Inform 6 `random` function rather than introducing a
+    new random function so that all randomness, even in extensions, uses the new
+    algorithm.
+    
+    But the I6 `random` function does more than just returning a random number.
+    As [documented in the DM4](https://www.inform-fiction.org/manual/html/s1.html#p45)
+    `random` can also be called with multiple arguments, for it to randomly choose
+    between. This however [does not work in Inform 7](https://intfiction.org/t/lack-of-randomness-sometimes-when-compiling-for-glulx/64533/101).
+    Inline invocations like this
+    ```
+    To decide which number is todays number: (- random(1,2,3,4,5,6,7,8,9) -).
+    ```
+    produce the error
+    > inform7: this inv of !random should have 1 argument(s), but has 9
+    
+    But calls to `random` in whole-function inclusions like this
+    ```
+    Include (-
+    [ TodaysNumber;
+        return random(1,2,3,4,5,6,7,8,9);
+    ];
+    -).
+    ```
+    compile, however they are silently modified during the Inter transformation
+    to only include one argument in the final build.
+
+    Inform 7 should consistently reject the multiple-argument form of `random`
+    at all times.
+
+4. There is a further complication: the Glulx I6 implementation of `random`
+    does not match the behaviour described in the DM4. Instead of negative or
+    zero values seeding the interpreter's RNG, it is passed directly to [the Glulx
+    `@random` opcode](https://www.eblong.com/zarf/glulx/Glulx-Spec.html#opcodes_rand),
+    before adding 1. This results in the following inconsistencies:
+
+    | Argument | DM4/Z-Machine | Glulx |
+    |----------|---------------|-------|
+    | Positive N | Return a random number in the range 1 to N | Return a random number in the range 1 to N |
+    | Zero | Switch the interpreter RNG to random mode | Return a full 32-bit random number |
+    | Negative N | Seed the interpreter RNG in predictable mode | Return a random number in the range (N + 2) to 1
+
+    Nobody would intentionally call `random(-10)` expecting it to return a number
+    between -8 and 1, so we don't need to be concerned with preserving the I6
+    behaviour of `random` in Glulx; instead we should make it conform to the DM4.
+
+5. People may find the behaviour of the Glulx `@random` opcode more useful than

```

Comment:
I'm not sure if it has to match `@random` exactly (its semantics are a bit awkward due to being essentially overloaded), but I'd definitely suggest an equivalent of `@random 0`, getting a full 32 bits of randomness. That's a very useful primitive for all sorts of things the standard rules don't and shouldn't provide, any underlying RNG can support it for free, and having to build it from the behavior of I6 `random()` or Glulx `@random nonZero` is possible but an unnecessary contortion.

Reply:
Matching the full `@random` opcode won't be difficult.

Reply:
It wouldn't be difficult, but would it be sensible? I don't know the last time I might have wanted what `@random n` does for negative n (and if I ever want it, it's easy to get that from the existing `random number between A and B` phrase). I can understand that a VM bytecode format wants to be frugal with opcode numbers and assign some possibly useful behavior to the entire range of inputs (here, one 32-bit integer). But there's no reason to replicate that idiosyncratic design 1:1 into Inform, where phrases are already plentiful and happily issue RTPs for nonsensical inputs. As an author, I might want "number from 0 to N exclusive" (e.g., as a list index) sometimes, and "arbitrary 32 bit number" at other times, but I never want to get one when I expected the other because a variable I expected to be positive turned out to be zero. I really don't see any reason to *not* make these two distinct phrases (and scrap the L1 < 0 case entirely).

Reply:
There's also no reason not to implement thr behaviour of the opcode. And just because you can't think of a reason why you wouldn't want the negative version doesn't mean someone else wouldn't. This point in the proposal is just to ensure that no one has reduced options compared to before. If someone was calling `@random` directly rather than the `random` function, they could replace their opcode invocation with a call `random_glulx` (if that's what we called it) to get the same result.